libs/api/domains/university-gateway/src/lib/universityGateway.service.ts

```diff
@@ -0,0 +1,221 @@
+        })),
+        modeOfDelivery: item.modeOfDelivery.map((m) =>
+          m.modeOfDelivery.toString(),
+        ),
+      })),
+    }
+  }
+
+  async getProgramById(
+    input: UniversityGatewayGetPogramInput,
+  ): Promise<UniversityGatewayProgramDetails> {
+    const res = await this.programApi.programControllerGetProgramDetails({
+      id: input.id,
+    })
+
+    const item = res.data
+
+    return {
+      id: item.id,
+      externalId: item.externalId,
+      active: item.active,
+      nameIs: item.nameIs,
+      nameEn: item.nameEn,
+      universityId: item.universityId,
+      universityContentfulKey: item.universityDetails.contentfulKey,
+      departmentNameIs: item.departmentNameIs,
+      departmentNameEn: item.departmentNameEn,
+      startingSemesterYear: item.startingSemesterYear,
+      startingSemesterSeason: item.startingSemesterSeason.toString(),
+      applicationStartDate: item.applicationStartDate,
+      applicationEndDate: item.applicationEndDate,
+      schoolAnswerDate: item.schoolAnswerDate,
+      studentAnswerDate: item.studentAnswerDate,
+      degreeType: item.degreeType.toString(),
+      degreeAbbreviation: item.degreeAbbreviation,
+      credits: item.credits,
+      descriptionIs: item.descriptionIs,
+      descriptionEn: item.descriptionEn,
+      durationInYears: item.durationInYears,
+      costPerYear: item.costPerYear,
+      iscedCode: item.iscedCode,
+      languages: item.languages,
+      searchKeywords: item.searchKeywords,
+      tag: item.tag.map((t) => ({
+        id: t.tagId,
+        code: t.details.code,
+        nameIs: t.details.nameIs,
+        nameEn: t.details.nameEn,
+      })),
+      modeOfDelivery: item.modeOfDelivery.map((m) =>
+        m.modeOfDelivery.toString(),
+      ),
+      externalUrlIs: item.externalUrlIs,
+      externalUrlEn: item.externalUrlEn,
+      admissionRequirementsIs: item.admissionRequirementsIs,
+      admissionRequirementsEn: item.admissionRequirementsEn,
+      studyRequirementsIs: item.studyRequirementsIs,
+      studyRequirementsEn: item.studyRequirementsEn,
+      costInformationIs: item.costInformationIs,
+      costInformationEn: item.costInformationEn,
+      courses: item.courses.map((c) => ({
+        id: c.details.id,
+        externalId: c.details.externalId,
+        nameIs: c.details.nameIs,
+        nameEn: c.details.nameEn,
+        credits: c.details.credits,
+        descriptionIs: c.details.descriptionIs,
+        descriptionEn: c.details.descriptionEn,
+        externalUrlIs: c.details.externalUrlIs,
+        externalUrlEn: c.details.externalUrlEn,
+        requirement: c.requirement.toString(),
+        semesterYear: c.semesterYear,
+        semesterYearNumber: c.semesterYear
+          ? item.startingSemesterYear - c.semesterYear + 1
+          : undefined,
+        semesterSeason: c.semesterSeason.toString(),
+      })),
+    }
+  }
+
+  async getUniversities(): Promise<UniversityGatewayUniversity[]> {
```

Comment:
Here is perfect to use a [`DataLoader`](https://github.com/island-is/island.is/tree/main/libs/nest/dataloader) pattern to populate the details from Contentful. DataLoader provides a bulk get and caching.

We used this in the GQL `auth` domain where you can find examples, where we were getting the organisation logo url. So there already exists a data loader in the `@island.is/cms` library. 

[Here](https://github.com/island-is/island.is/blob/main/libs/api/domains/auth/src/lib/resolvers/domain.resolver.ts) you can find a example of how we created a fieldResolver for the `organisationLogoUrl` and are using the data loader to resolve the field value.